akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala

```diff
@@ -324,19 +325,21 @@ private[client] object NewHostConnectionPool {
 
           def settings: ConnectionPoolSettings = _settings
 
-          def openConnection(): Future[Http.OutgoingConnection] = {
+          def openConnection(): Unit = {
             if (connection ne null) throw new IllegalStateException("Cannot open connection when slot still has an open connection")
 
             connection = logic.openConnection(this)
-            connection.outgoingConnection
           }
           def pushRequestToConnectionAndThen(request: HttpRequest, nextState: SlotState): SlotState = {
             if (connection eq null) throw new IllegalStateException("Cannot open push request to connection when there's no connection")
 
-            // bit of a HACK to make sure onRequestEntityCompleted will end up in the right place
-            state = nextState
+            // bit of a HACK: pushing the request may cause a onRequestEntityCompleted event on the current thread.
 
+            // To accomodate this we first do an 'early' update of the state:
+            state = nextState
+            // Then execute the action that might cause the 'inline' state change:
             connection.pushRequest(request)
+            // And then return the possibly-again-updated state so we don't overwrite it afterwards:
```

Comment:
thanks for the comments, good to make it more understandable while we're here ;)

Reply:
yeah this had me puzzled for a while ;)